src/hotspot/share/opto/regmask.hpp

```diff
@@ -64,28 +75,300 @@ class RegMask {
+
   union {
-    // Array of Register Mask bits.  This array is large enough to cover
-    // all the machine registers and all parameters that need to be passed
-    // on the stack (stack registers) up to some interesting limit.  Methods
-    // that need more parameters will NOT be compiled.  On Intel, the limit
-    // is something like 90+ parameters.
+    // Array of Register Mask bits.  This array is large enough to cover all
+    // the machine registers and usually all parameters that need to be passed
+    // on the stack (stack registers) up to some interesting limit. On Intel,
+    // the limit is something like 90+ parameters.
+
+    // Viewed as an array of 32-bit words
     int       _RM_I[RM_SIZE];
+
+    // Viewed as an array of machine words
     uintptr_t _RM_UP[_RM_SIZE];
   };
 
-  // The low and high water marks represents the lowest and highest word
-  // that might contain set register mask bits, respectively. We guarantee
-  // that there are no bits in words outside this range, but any word at
-  // and between the two marks can still be 0.
+  // In rare situations (e.g., "more than 90+ parameters on Intel"), we need to
+  // extend the register mask with dynamically allocated memory. We keep the
+  // base statically allocated _RM_UP, and arena allocate the extended mask
+  // (RM_UP_EXT) separately. Another, perhaps more elegant, option would be to
+  // have two subclasses of RegMask, where one is statically allocated and one
+  // is (entirely) dynamically allocated. Given that register mask extension is
+  // rare, we decided to use the current approach (_RM_UP and _RM_UP_EXT) to
+  // keep the common case fast. Most of the time, we will then not need to
+  // dynamically allocate anything.
+  //
+  // We could use a GrowableArray here, but there are currently some
+  // GrowableArray limitations that have a negative performance impact for our
+  // use case:
+  //
+  // - There is no efficient copy/clone operation.
+  // - GrowableArray construction currently default-initializes everything
+  //   within the array's initial capacity, which is unnecessary in our case.
+  //
+  // After addressing these limitations, we should consider using a
+  // GrowableArray here.
+  uintptr_t* _RM_UP_EXT = nullptr;
+
+#ifdef ASSERT
+  // Register masks may get shallowly copied without the use of constructors,
+  // which is problematic when dealing with the externally allocated memory for
+  // _RM_UP_EXT. Therefore, we need some sanity checks to ensure we have not
+  // missed any such cases. The below variables enable such checks.
+  //
+  // The original address of the _RM_UP_EXT variable, set when using
+  // constructors. If we get copied/cloned, &_RM_UP_EXT will no longer equal
+  // _original_ext_address.
+  uintptr_t** _original_ext_address = &_RM_UP_EXT;
+  //
+  // If the original version, of which we may be a clone, is read-only. In such
+  // cases, we can allow read-only sharing.
+  bool _read_only = false;
+#endif
```

Comment:
Can you explain why this happens? Is this something we could clean up? It smells a bit like tech-dept. But maybe it is a really necessary performance optimization. Would be nice if there was an explanation which one it is ;)

Reply:
Sure, will do.